### ISSUE TITLE
Sprite animation timings?

### ISSUE THREAD

**Issue by johanhelsing** [KirmesBude/bevy_titan#2]
Hey, I've been looking for something like this to improve my Krita -> bevy workflow.

I'm using Krita to do sprite animations, and currently I'm animating at 60 FPS, because that's what my game logic and networking solution runs at. However, when holding a pose for several frames (which happens a lot due to the high frame rate), I don't want to have duplicate textures.

So I'm looking for a system that will also allow me to also embed some timing information in the manifest.

If this is out of scope for the project, I'm perfectly happy forking (thanks for sharing!), but I thought I'd ask in case this is something you'd want to support as well.

**Comment by KirmesBude** [comment]
I have https://github.com/KirmesBude/bevy_trickfilm that allows you to define keyframe (as index) and keyframe time. 
Maybe that is what you are looking for?

**Comment by johanhelsing** [comment]
Yes, that's almost exactly what I want! I'd prefer something that operates on frames instead of time, since I need rollback and manual stepping to work, and I'll likely have actions and hitboxes tied to specific frames, but perhaps I can make it work. If not, I can probably patch it to make it to what I want.

Nevertheless, I'm pretty sure I can use bevy_titan unmodified. It's probably a better idea to add this stuff on top like you did in trickfilm anyway.

**Comment by johanhelsing** [comment]
Alright, sharing some of my findings here: https://github.com/johanhelsing/krita_bevy_anim

Getting closer to a working pipeline, hopefully something that might be useful for others too at some point.

**Comment by johanhelsing** [comment]
I ended up using my own frame-based animation manifest format, but adding support for bevy_trickfilm manifest to that pipeline as well should be relatively easy.

**Comment by KirmesBude** [comment]
It looks like your "Animation Asset" does not have any reference to the Image data. 
I have been thinking about something similar for https://github.com/KirmesBude/bevy_trickfilm/issues/8. 

How do you plan to handle playback? Would you expect the user to give both a Flippy handle and a TextureAtlas handle to your version of an AnimationPlayer - or would you leave the second part up to usercode?

**Comment by johanhelsing** [comment]
I just require files to be named the same, so `player_run.flippy` always tries to load `player_run.titan`. It won't let me have multiple animations per sprite sheet, but that would also be pretty hard anyway, given how Krita works.

Some snippets of the code I use at the moment:

```rust
let clip = anim_clips.get(&assets.run)?;
sprite.index = clip.get_index(frame);
```

```rust
impl AnimClip {
    pub(crate) fn get_index(&self, frame: usize) -> usize {
        let frame = frame % self.length as usize; // wrap around
        self.timings
            .iter()
            .rposition(|t| (*t as usize) <= frame)
            .unwrap_or(0)
    }
}
```

```rust
impl AssetLoader for FlippyLoader {
    fn load<'a>(
        &'a self,
        bytes: &'a [u8],
        load_context: &'a mut LoadContext,
    ) -> bevy::utils::BoxedFuture<'a, Result<(), bevy::asset::Error>> {
        Box::pin(async move {
            let Flippy { length, timings } = ron::de::from_bytes::<Flippy>(bytes)?;

            let titan_path = {
                let mut path = PathBuf::from(load_context.path());
                path.set_extension("titan");
                path
            };

            let atlas_path = AssetPath::new_ref(titan_path.as_path(), None);
            let atlas: Handle<TextureAtlas> = load_context.get_handle(atlas_path.clone());

            let anim_clip = AnimClip {
                atlas,
                timings,
                length,
            };

            let anim_clip_asset = LoadedAsset::new(anim_clip).with_dependency(atlas_path);
            load_context.set_default_asset(anim_clip_asset);
            Ok(())
        })
    }
```

**Comment by KirmesBude** [comment]
Oh, that's smart. 
Cool stuff.
